scrapy/core/spidermw.py

```diff
@@ -69,47 +86,43 @@ def process_spider_exception(_failure, start_index=0):
                 elif result is None:
                     continue
                 else:
-                    raise _InvalidOutput('Middleware {} must return None or an iterable, got {}'
-                                         .format(fname(method), type(result)))
+                    msg = "Middleware {} must return None or an iterable, got {}"
+                    raise _InvalidOutput(msg.format(_fname(method), type(result)))
             return _failure
 
         def process_spider_output(result, start_index=0):
             # items in this iterable do not need to go through the process_spider_output
             # chain, they went through it already from the process_spider_exception method
             recovered = MutableChain()
 
-            def evaluate_iterable(iterable, index):
-                try:
-                    for r in iterable:
-                        yield r
-                except Exception as ex:
-                    exception_result = process_spider_exception(Failure(ex), index+1)
-                    if isinstance(exception_result, Failure):
-                        raise
-                    recovered.extend(exception_result)
-
             method_list = islice(self.methods['process_spider_output'], start_index, None)
             for method_index, method in enumerate(method_list, start=start_index):
                 if method is None:
                     continue
-                # the following might fail directly if the output value is not a generator
                 try:
+                    # might fail directly if the output value is not a generator
                     result = method(response=response, result=result, spider=spider)
                 except Exception as ex:
                     exception_result = process_spider_exception(Failure(ex), method_index+1)
                     if isinstance(exception_result, Failure):
                         raise
                     return exception_result
-                if _isiterable(result):
-                    result = evaluate_iterable(result, method_index)
                 else:
```

Comment:
💄 is else needed here?

Reply:
Not really, because of the above `return exception_result`. I just thought it would make the separation clearer.

Reply:
Removed in ad4477d335bee8b10bc3bbca969defddd9b316f8.
The `else` is still there and the comment does not appear as "outdated" because it's now part of the `if _isiterable(result)` block, not the `try...except` one.